Ubuntu 22.04 compilation issue: "error: ‘O_CLOEXEC’ undeclared (first use in this function); did you mean ‘FD_CLOEXEC’?"

I tried gcc 10,11 and 12 with the same result.

Repro on 6.5.0-41-generic #41~22.04.2-Ubuntu :

1. `meson setup build`
2. `meson compile` or `ninja` within the `\build` folder:

```
[2/119] Compiling C object deps/hidapi/linux/libhidapi-hidraw.a.p/hid.c.o
FAILED: deps/hidapi/linux/libhidapi-hidraw.a.p/hid.c.o
gcc-12 -Ideps/hidapi/linux/libhidapi-hidraw.a.p -Ideps/hidapi/linux -I../deps/hidapi/linux -Ideps/hidapi/hidapi -I../deps/hidapi/hidapi -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c11 -Os -g -fPIC -pthread -MD -MQ deps/hidapi/linux/libhidapi-hidraw.a.p/hid.c.o -MF deps/hidapi/linux/libhidapi-hidraw.a.p/hid.c.o.d -o deps/hidapi/linux/libhidapi-hidraw.a.p/hid.c.o -c ../deps/hidapi/linux/hid.c
../deps/hidapi/linux/hid.c: In function ‘utf8_to_wchar_t’:
../deps/hidapi/linux/hid.c:111:32: warning: implicit declaration of function ‘wcsdup’; did you mean ‘wcscmp’? [-Wimplicit-function-declaration]
  111 |                         return wcsdup(L"");
      |                                ^~~~~~
      |                                wcscmp
../deps/hidapi/linux/hid.c:111:32: warning: returning ‘int’ from a function with return type ‘wchar_t *’ {aka ‘int *’} makes pointer from integer without a cast [-Wint-conversion]
  111 |                         return wcsdup(L"");
      |                                ^~~~~~~~~~~
../deps/hidapi/linux/hid.c: In function ‘get_hid_report_descriptor’:
../deps/hidapi/linux/hid.c:361:48: error: ‘O_CLOEXEC’ undeclared (first use in this function); did you mean ‘FD_CLOEXEC’?
  361 |         rpt_handle = open(rpt_path, O_RDONLY | O_CLOEXEC);
      |                                                ^~~~~~~~~
      |                                                FD_CLOEXEC
../deps/hidapi/linux/hid.c:361:48: note: each undeclared identifier is reported only once for each function it appears in
../deps/hidapi/linux/hid.c: In function ‘parse_hid_vid_pid_from_uevent_path’:
../deps/hidapi/linux/hid.c:450:47: error: ‘O_CLOEXEC’ undeclared (first use in this function); did you mean ‘FD_CLOEXEC’?
  450 |         handle = open(uevent_path, O_RDONLY | O_CLOEXEC);
      |                                               ^~~~~~~~~
      |                                               FD_CLOEXEC
../deps/hidapi/linux/hid.c: In function ‘parse_uevent_info’:
../deps/hidapi/linux/hid.c:552:46: warning: implicit declaration of function ‘strdup’; did you mean ‘strcmp’? [-Wimplicit-function-declaration]
  552 |                         *product_name_utf8 = strdup(value);
      |                                              ^~~~~~
      |                                              strcmp
../deps/hidapi/linux/hid.c:552:44: warning: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  552 |                         *product_name_utf8 = strdup(value);
      |                                            ^
../deps/hidapi/linux/hid.c:556:45: warning: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  556 |                         *serial_number_utf8 = strdup(value);
      |                                             ^
../deps/hidapi/linux/hid.c: In function ‘create_device_info_for_device’:
../deps/hidapi/linux/hid.c:634:51: warning: pointer/integer type mismatch in conditional expression
  634 |         cur_dev->path = dev_path? strdup(dev_path): NULL;
      |                                                   ^
../deps/hidapi/linux/hid.c:667:62: warning: assignment to ‘wchar_t *’ {aka ‘int *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  667 |                                 cur_dev->manufacturer_string = wcsdup(L"");
      |                                                              ^
../deps/hidapi/linux/hid.c:693:54: warning: assignment to ‘wchar_t *’ {aka ‘int *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  693 |                         cur_dev->manufacturer_string = wcsdup(L"");
      |                                                      ^
../deps/hidapi/linux/hid.c:700:54: warning: assignment to ‘wchar_t *’ {aka ‘int *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  700 |                         cur_dev->manufacturer_string = wcsdup(L"");
      |                                                      ^
../deps/hidapi/linux/hid.c:708:54: warning: assignment to ‘wchar_t *’ {aka ‘int *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  708 |                         cur_dev->manufacturer_string = wcsdup(L"");
      |                                                      ^
../deps/hidapi/linux/hid.c:750:67: warning: pointer/integer type mismatch in conditional expression
  750 |                         cur_dev->path = dev_path? strdup(dev_path): NULL;
      |                                                                   ^
../deps/hidapi/linux/hid.c:753:106: warning: pointer/integer type mismatch in conditional expression
  753 |                         cur_dev->serial_number = prev_dev->serial_number? wcsdup(prev_dev->serial_number): NULL;
      |                                                                                                          ^
../deps/hidapi/linux/hid.c:756:124: warning: pointer/integer type mismatch in conditional expression
  756 |                         cur_dev->manufacturer_string = prev_dev->manufacturer_string? wcsdup(prev_dev->manufacturer_string): NULL;
      |                                                                                                                            ^
../deps/hidapi/linux/hid.c:757:109: warning: pointer/integer type mismatch in conditional expression
  757 |                         cur_dev->product_string = prev_dev->product_string? wcsdup(prev_dev->product_string): NULL;
      |                                                                                                             ^
../deps/hidapi/linux/hid.c: In function ‘hid_open_path’:
../deps/hidapi/linux/hid.c:1001:50: error: ‘O_CLOEXEC’ undeclared (first use in this function); did you mean ‘FD_CLOEXEC’?
 1001 |         dev->device_handle = open(path, O_RDWR | O_CLOEXEC);
      |                                                  ^~~~~~~~~
      |                                                  FD_CLOEXEC
[3/119] Generating version.h with a custom command
ninja: build stopped: subcommand failed.
```

--- a/README.md
+++ b/README.md
@@ -132,7 +132,7 @@ use the `/dev/tty.usbmodemXXX1` device node involving how it operates under the 
 When developing this project, the following tools are required:
 
 * [git](https://git-scm.com)
-* [meson](https://mesonbuild.com/Getting-meson.html) (version 0.61 or greater)
+* [meson](https://mesonbuild.com/Getting-meson.html) (version 0.63 or greater)
 * [ninja](https://ninja-build.org)
 
 Additionally, depending on if you want to build/work on the firmware or Black Magic Debug App (BMDA), you also need:

--- a/README.md
+++ b/README.md
@@ -12,12 +12,14 @@ The project allows debugging of devices connected over JTAG or SWD, and via the 
 [bmpflash](https://github.com/blackmagic-debug/bmpflash) the programming of SPI Flash devices.
 This includes support for ARM and RISC-V devices, the complete list can be found on the website.
 
+[![Discord](https://img.shields.io/discord/613131135903596547?logo=discord)](https://discord.gg/P7FYThy)
+[![Current release](https://img.shields.io/github/v/release/blackmagic-debug/blackmagic.svg?logo=github)](https://github.com/blackmagic-debug/blackmagic/releases)
+[![CI flow status](https://github.com/blackmagic-debug/blackmagic/actions/workflows/build-and-upload.yml/badge.svg)](https://github.com/blackmagic-debug/blackmagic/actions/workflows/build-and-upload.yml)
+
 ## Resources
 
 * [Official website](https://black-magic.org/index.html)
 * [Binary builds](https://github.com/blackmagic-debug/blackmagic/releases)
-
-[![Discord](https://img.shields.io/discord/613131135903596547?logo=discord)](https://discord.gg/P7FYThy)
 
 ## Usage
 

--- a/README.md
+++ b/README.md
@@ -35,16 +35,18 @@ however below is a brief guide for both the firmware and BMDA.
 
 ### Black Magic Debug Firmware
 
+When built as firmware and put on a probe, the project is used like as follows:
+
 ```console
 > arm-none-eabi-gdb gpio.elf
 ...<GDB Copyright message>
 (gdb) tar ext /dev/ttyBmpGdb
 Remote debugging using /dev/ttyBmpGdb
-(gdb) mon s
+(gdb) mon a
 Target voltage: 2.94V
 Available Targets:
 No. Att Driver
- 1      STM32F40x M3/M4
+ 1      STM32F40x M4
 (gdb) att 1
 Attaching to program: /devel/en_apps/gpio/f4_discovery/gpio.elf, Remote target
 0x08002298 in UsartIOCtl ()
@@ -65,17 +67,20 @@ Breakpoint 1, main () at /devel/en_apps/gpio/f4_discovery/../gpio.c:70
 (gdb)
 ```
 
+Note: this presumes on Linux that you have installed the project's udev rules.
+
 ### Black Magic Debug App
 
-You can also build the Black Magic Debug suite as a PC program called Black Magic Debug App (BMDA).
+When built as BMDA, usage differs a bit - principally as instead of using a virtual serial interface,
+BMDA makes the GDB server available over TCP/IP on a port that it will print to its console (typically
+2000 though). This means replacing `tar ext /dev/ttyBmpGdb` in the firmware example, with `tar ext :2000`.
 
-This builds the same GDB server that is running on the Black Magic Probe.
-While connection to the Black Magic Probe GDB server is via serial line,
-connection to the Black Magic Debug App is via TCP port 2000 for the first
-GDB server and higher for more invocations. Use "tar(get) ext(ented) :2000"
-to connect.
+To bring the GDB server up, you will need a probe it can talk to, the BMDA binary, and an extra console to
+the one you want to run GDB from. Verbosity of BMDA can be increased with the `-v` option such as running
+`blackmagic -v 5` which sets the server to emit INFO and TARGET diagnostic level information in addition to
+ERROR and WARNING output.
 
-Black Magic Debug App can talk to
+Black Magic Debug App can talk to:
 
 * Black Magic Probe firmware probes via the USB-serial port
 * ST-Link v2, v2.1 and v3 with recent firmware
@@ -83,40 +88,42 @@ Black Magic Debug App can talk to
 * J-Link probes
 * FTDI MPSSE based probes
 
-When connected to a single BMP supported probe, starting `blackmagic` w/o any
-arguments starts the server. When several BMP supported probes are connected,
-their types, position and serial number is displayed and the program exits.
-Add `-P (position)` to the next invocation to select one.
+When connected to a single BMDA-supported probe, starting `blackmagic` w/o any arguments starts the server.
+When several BMDA-supported probes are connected, their types, serial numbers and version information is
+displayed and the program exits. Add `-s <serial>` to the next invocation to select one.
+
 For the setup from the sample session above:
 
 In another terminal:
 
 ```console
 > blackmagic
-Black Magic Debug App v1.9.2
+Black Magic Debug App v1.10.2
  for Black Magic Probe, ST-Link v2 and v3, CMSIS-DAP, J-Link and FTDI (MPSSE)
 Using 1d50:6018 8BB20A03 Black Magic Debug
- Black Magic Probe  v1.9.2
+ Black Magic Probe v1.10.2
+Setting V6ONLY to off for dual stack listening.
 Listening on TCP: 2000
 ```
 
 And in the GDB terminal:
 
 ```console
+...
 (gdb) tar ext :2000
 Remote debugging using :2000
-(gdb) mon s
+(gdb) mon a
 ...
 ```
 
 Black Magic Debug App also provides for Flashing, reading and verification of a binary file,
-by default starting at lowest flash address. The `-t` argument displays information about the
+by default starting at lowest Flash address. The `-t` argument displays information about the
 connected target. Use `-h`/`--help` to get a list of supported options.
 
 #### OS specific remarks
 
-On *BSD and macOS, you should use `/dev/cu.usbmodemXXX`. There are unresolved issues with trying to
-use the `/dev/tty.usbmodemXXX` device node involving how it operates under the hood.
+On *BSD and macOS, you should use `/dev/cu.usbmodemXXX1`. There are unresolved issues with trying to
+use the `/dev/tty.usbmodemXXX1` device node involving how it operates under the hood.
 
 ## Getting started
 

--- a/README.md
+++ b/README.md
@@ -1,18 +1,16 @@
-# Black Magic Probe
+# Black Magic Debug
 
-Firmware for the Black Magic Debug Probe.
+The Black Magic Debug suite is a self-contained debugger for microcontrollers which needs no config
+files and auto-detects plus configures the connected targets. It is designed to be fast and easy to use.
 
-The Black Magic Probe is a modern, in-application debugging tool for
-embedded microprocessors. It allows you see what is going on 'inside' an
-application running on an embedded microprocessor while it executes. It is
-able to control and examine the state of the target microprocessor using a
-JTAG or Serial Wire Debugging (SWD) port and on-chip debug logic provided
-by the microprocessor. The probe connects to a host computer using a
-standard USB interface. The user is able to control exactly what happens
-using the GNU source level debugging software, GDB.
-Serial Wire Output (SWO) allows the target to write tracing and logging to the host
-without using usb or serial port. Decoding SWO in the probe itself
-makes [SWO viewing as simple as connecting to a serial port](https://black-magic.org/usage/swo.html).
+The project can be built as either firmware for
+[Black Magic Probe](https://1bitsquared.com/products/black-magic-probe) -
+a debugger-in-a-dongle that provides multi-voltage debug with no other external tools than GDB required -
+or as Black Magic Debug App which is the project built for the host machine, more details below.
+
+The project allows debugging of devices connected over JTAG or SWD, and via the companion tool
+[bmpflash](https://github.com/blackmagic-debug/bmpflash) the programming of SPI Flash devices.
+This includes support for ARM and RISC-V devices, the complete list can be found on the website.
 
 ## Resources
 
@@ -28,8 +26,8 @@ makes [SWO viewing as simple as connecting to a serial port](https://black-magic
 ```console
 > arm-none-eabi-gdb gpio.elf
 ...<GDB Copyright message>
-(gdb) tar ext /dev/ttyACM0
-Remote debugging using /dev/ttyACM0
+(gdb) tar ext /dev/ttyBmpGdb
+Remote debugging using /dev/ttyBmpGdb
 (gdb) mon s
 Target voltage: 2.94V
 Available Targets:

--- a/README.md
+++ b/README.md
@@ -275,7 +275,8 @@ You should now see the resulting executable in `build`:
 
 ### Changing the build configuration
 
-You may change the configuration at any time after configuring the project, if may also change the default options while configuring for the first time.
+You may change the configuration at any time after configuring the project, if may also change the default
+options while configuring for the first time.
 
 Changing options at configure time:
 
@@ -295,17 +296,25 @@ Alternatively, you can use the equivalent but more verbose `meson setup` command
 meson setup build --reconfigure -Dtargets=cortexm,stm
 ```
 
-Keep in mind that when changing the configured probe, the other default options will not change, and you may end up with a configuration that does not make sense for it. In such cases it's best to use the *cross-file* for the probe, not just change the `probe` option. For this you will need to use the `meson setup` command again:
+NB: Since about Meson v1.2.0, using this second form is the only way to change the active targets list. This is
+a known issue in Meson which has yet to be addressed in a release.
+
+Keep in mind that when changing the configured probe, the other default options will not change, and you may end
+up with a configuration that does not make sense for it. In such cases it's best to use the *cross-file* for the
+probe, not just change the `probe` option. For this you will need to use the `meson setup` command again:
 
 ```sh
 meson setup build --reconfigure --cross-file=cross-file/bluepill.ini
 ```
 
-When changing options after configuration, you may omit the argument `build` if you are running the command from within the `build` directory.
+When changing options after configuration, you may omit the argument `build` if you are running the command from
+within the `build` directory.
 
-You can have multiple build directories! So if you are regularly building firmware for multiple probes we would recommend keeping an individual build directory configured for each one.
+You can have multiple build directories! So if you are regularly building firmware for multiple probes we would
+recommend keeping an individual build directory configured for each one.
 
-If you are working with PowerShell you may have some issue while trying to configure some options like the enabled target list `-Dtargets=cortexm,stm`:
+If you are working with PowerShell you may have some issue while trying to configure some options like the enabled
+target list `-Dtargets=cortexm,stm`:
 
 ``` console
 PS C:\...\blackmagic\build> meson configure build -Dtargets=cortexm,stm

--- a/README.md
+++ b/README.md
@@ -163,8 +163,8 @@ do source builds of the dependencies using the information in `deps/` that will 
 
 ### Building
 
-The project is configured and built using the Meson build system, you will need to create a build directory
-and then configure the build depending on what you want.
+The project is configured and built using the Meson build system, you will need to create a build
+directory using `meson setup`, and configure the build depending on what you want.
 
 The project has a couple of build options you may configure, which affect the resulting binaries.
 This is how you will configure the firmware for your respective probe, or enable certain features.
@@ -172,11 +172,12 @@ This is how you will configure the firmware for your respective probe, or enable
 A non-exhaustive list of project options:
 
 * `probe`: Hardware platform where the BMD firmware will run
-* `targets`: Enabled debug targets
+* `targets`: Targets and architectures enabled for debug and Flashing
 * `debug_output`: Enable debug output (for debugging the BMD stack, not debug targets)
 * `rtt_support`: Enable RTT (Real Time Transfer) support
 
-You may see all available project options and valid values under  `Project options` in the output of the `meson configure` command.
+You may see all available project options and valid values under `Project options` in the output
+of the `meson configure` command.
 
 The following commands are ran from the root of your clone of the `blackmagic` repository:
 
@@ -184,11 +185,13 @@ The following commands are ran from the root of your clone of the `blackmagic` r
 cd /path/to/repository/blackmagic
 ```
 
-#### Building Black Magic Debug Firmware
+#### Building the firmware
 
-To build the firmware you need to configure the probe hardware the firmware will run on, as well as the cross-compilation toolchain to be used.
+To build the firmware you need to configure the probe hardware the firmware will run on, as well as the
+cross-compilation toolchain to be used.
 
-For convenience, a cross-file for each supported hardware probe is available which provides a sane default configuration for it.
+For convenience, a cross-file for each supported hardware probe is available which provides a sane default
+configuration for it.
 
 The build configuration command for the native probe may look like:
 
@@ -196,17 +199,22 @@ The build configuration command for the native probe may look like:
 meson setup build --cross-file cross-file/native.ini
 ```
 
-Note that even if you are using the pre-configured cross-file, you may still override it's defaults with `-Doption=value` in the same configuration command, or later as highlighted in [Changing the build configuration](#changing-the-build-configuration).
+Note that even if you are using the pre-configured cross-file, you may still override it's defaults with
+`-Doption=value` in the same configuration command, or later as highlighted in
+[Changing the build configuration](#changing-the-build-configuration).
 
-Alternatively (for advanced users), if you wish to configure manually, for instance while writing support for a new probe, or a different toolchain, you can run:
+Alternatively (for advanced users), if you wish to configure manually, for instance while writing support
+for a new probe, or a different toolchain, you can run something similar to this:
 
 ```sh
 meson setup build --cross-file cross-file/arm-none-eabi.ini -Dprobe=native -Dtargets=cortexm,stm
 ```
 
-You now should have a `build` directory from where you can build the firmware, this is also where your binary files will appear.
+After following one of these two paths, you now should have a `build` directory from where you can build
+the firmware. This is also where your binary files will appear.
 
-The command `meson compile` will build the default targets, you may omit `-C build` if you run the command from within the `build` directory:
+The command `meson compile` will build the default targets, you may omit `-C build` if you run the command
+from within the `build` directory:
 
 ```sh
 meson compile -C build
@@ -221,7 +229,8 @@ These are the binary files you will use to flash to your probe.
 
 ##### region `rom' overflowed
 
-It may happen, while working with non default configurations or the project's latest version from Git, that the firmware does not fit in the available space for the configured probe, this could look something like:
+It may happen, while working with non default configurations or the project's latest version from Git,
+that the firmware does not fit in the available space for the configured probe, this could look something like:
 
 ```console
 arm-none-eabi/bin/ld: region `rom' overflowed by 4088 bytes
@@ -231,7 +240,8 @@ Memory region         Used Size  Region Size  %age Used
 collect2: error: ld returned 1 exit status
 ```
 
-This is not unexpected, as some hardware probe have limited space for firmware. You can get around it by disabling some features or targets support:
+This is not unexpected, as some hardware probe have limited space for firmware. You can get around it by
+disabling some features or targets support:
 
 ```sh
 meson configure build -Dtargets=cortexm,stm -Drtt_support=false
@@ -239,17 +249,21 @@ meson configure build -Dtargets=cortexm,stm -Drtt_support=false
 
 #### Building Black Magic Debug App
 
-The Black Magic Debug App (BMDA) is always built by default, even for firmware builds. So long as all its dependencies are found, you can find the executable under the `build` directory, named simply `blackmagic`.
+The Black Magic Debug App (BMDA) is always built by default, even for firmware builds. So long as all its
+dependencies are found, you can find the executable under the `build` directory, named simply `blackmagic`.
 
-If you wish to build only BMDA, you can set the hardware `probe` option to an empty string `-Dprobe=''`, this is the default value:
+If you wish to build only BMDA, you can set the hardware `probe` option to an empty string `-Dprobe=''`,
+this is the default value:
 
 ```sh
 meson setup build
 ```
 
-You now should have a `build` directory from where you can build the app, this is also where your executable will appear.
+You now should have a `build` directory from where you can build the app, this is also where your executable
+will appear.
 
-The command `meson compile` will build the default targets, you may omit `-C build` if you run the command from within the `build` directory:
+The command `meson compile` will build the default targets, you may omit `-C build` if you run the command
+from within the `build` directory:
 
 ```sh
 meson compile -C build

--- a/README.md
+++ b/README.md
@@ -137,25 +137,29 @@ When developing this project, the following tools are required:
 
 Additionally, depending on if you want to build/work on the firmware or Black Magic Debug App (BMDA), you also need:
 
-#### Black Magic Debug Firmware
+#### Black Magic Debug Firmware Requirements
 
 * [`arm-none-eabi-gcc`](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/downloads) (from ARM, self-built and distro-supplied is currently broken)
 
-If you wish to use the older [gnu-rm](https://developer.arm.com/downloads/-/gnu-rm) ARM toolchain,
-this is fine and works well.
+The project is tested to work on the 12.2.Rel1 and 13.2.Rel1 releases. Newer releases may work but are not guaranteed.
+The old `gnu-rm` compiler series is no longer supported as the code generation performs way too badly.
 
 If you have a toolchain from other sources and find problems, check if it's an issue with your toolchain first,
 and if not then open an issue, or better, submit a pull request with a fix.
 
-#### Black Magic Debug App
+#### Black Magic Debug App Requirements
 
-* [GCC](https://gcc.gnu.org) or Clang (Clang is not officially supported)
-
-If you wish to enable support for 3rd party probes, and not only the official native hardware, you will also need:
-
+* [GCC](https://gcc.gnu.org) or Clang
+* pkg-config
 * libusb1
-* libftdi1
 * libhidapi
+
+On Linux and macOS (+ other non-Windows platforms):
+
+* libftdi1
+
+If you do not install the subsequent dependencies libusb1, libhidapi, and libftdi1, then the build system will
+do source builds of the dependencies using the information in `deps/` that will be statically linked into BMDA.
 
 ### Building
 

--- a/README.md
+++ b/README.md
@@ -6,7 +6,7 @@ files and auto-detects plus configures the connected targets. It is designed to 
 The project can be built as either firmware for
 [Black Magic Probe](https://1bitsquared.com/products/black-magic-probe) -
 a debugger-in-a-dongle that provides multi-voltage debug with no other external tools than GDB required -
-or as Black Magic Debug App which is the project built for the host machine, more details below.
+or as Black Magic Debug App (BMDA) which is the project built for the host machine, more details below.
 
 The project allows debugging of devices connected over JTAG or SWD, and via the companion tool
 [bmpflash](https://github.com/blackmagic-debug/bmpflash) the programming of SPI Flash devices.
@@ -16,12 +16,22 @@ This includes support for ARM and RISC-V devices, the complete list can be found
 [![Current release](https://img.shields.io/github/v/release/blackmagic-debug/blackmagic.svg?logo=github)](https://github.com/blackmagic-debug/blackmagic/releases)
 [![CI flow status](https://github.com/blackmagic-debug/blackmagic/actions/workflows/build-and-upload.yml/badge.svg)](https://github.com/blackmagic-debug/blackmagic/actions/workflows/build-and-upload.yml)
 
+Table of contents:
+
+* [Resources](#resources)
+* [Usage](#usage)
+* [Build quick-start](#getting-started)
+* [Contribution information](#contributing-and-reporting-issues)
+
 ## Resources
 
 * [Official website](https://black-magic.org/index.html)
 * [Binary builds](https://github.com/blackmagic-debug/blackmagic/releases)
 
 ## Usage
+
+There is a more detailed [getting started guide](https://black-magic.org/getting-started.html) on the website,
+however below is a brief guide for both the firmware and BMDA.
 
 ### Black Magic Debug Firmware
 
@@ -52,6 +62,7 @@ Note: automatically using hardware breakpoints for read-only addresses.
 
 Breakpoint 1, main () at /devel/en_apps/gpio/f4_discovery/../gpio.c:70
 70      {
+(gdb)
 ```
 
 ### Black Magic Debug App

--- a/README.md
+++ b/README.md
@@ -330,32 +330,19 @@ To get around this you may wrap the options with double quotes `"`, in this exam
 meson configure build "-Dtargets=cortexm,stm"
 ```
 
-### Working with an existing clone (used before the new meson build system was introduced)
+### Working with an existing clone
 
-If you are working with an existing clone of the project where you used the old `make` build system,
-or simply initialized the submodules, you may encounter issues while trying the new Meson workflow,
-this may look something like:
+If you are working with an existing clone of the project, you may encounter a build issue with libopencm3 or
+one of the other dependencies in `deps/` - if this happens, you have one of two recourses:
 
-```console
-fatal: Fetched in submodule path 'deps/libopencm3', but it did not contain bed4a785eecb6c9e77e7491e196565feb96c617b. Direct fetching of that commit failed.
-```
+The first is to ensure the dependency is fully up to date (Meson manages them as Git clones but due to limitations
+is unaware of when the remote is newer than the local when using `revision = head` in the .wrap files) with the
+remote by doing a full fetch cycle (`git fetch --all --prune`, `git pull --tags --force`) within the dependency
+clone, and making sure the latest version is checked out (`git pull`). This is labour intensive but good if
+you have working copy changes
 
-To get around this try running the following commands:
-
-```sh
-cd deps/libopencm3
-git remote set-url origin https://github.com/blackmagic-debug/libopencm3
-cd ../..
-git submodule update
-```
-
-Alternatively, the nuclear option
-(**BEWARE, THIS WILL ERASE ANY CHANGES THAT HAVE NOT BEEN PUSHED UPSTREAM**)
-
-```sh
-git submodule deinit --force --all
-meson subprojects purge --confirm
-```
+The second, more nuclear option, is to blow away the clone of the dependency, eg `rm -rf deps/libopencm3` and rebuild.
+Meson will automatically detect this and re-clone the dependency at the version indicated by the .wrap files.
 
 ## Contributing and reporting issues
 

--- a/meson.build
+++ b/meson.build
@@ -43,7 +43,7 @@ project(
 		'werror=false',
 		'b_ndebug=if-release',
 	],
-	meson_version: '>= 0.61.0',
+	meson_version: '>= 0.63.0',
 	subproject_dir: 'deps',
 )
 

